video/out/wayland_common.c

```diff
@@ -1201,6 +1182,34 @@ static struct vo_wayland_output *find_output(struct vo_wayland_state *wl, int in
     return NULL;
 }
 
+void vo_wayland_wait_frame_callback(struct vo_wayland_state *wl)
+{
+    struct pollfd fds[1] = {
+        {.fd = wl->display_fd,     .events = POLLIN },
+    };
+
+    double vblank_time = 1e6 / wl->current_output->refresh_rate;
+    int64_t finish_time = mp_time_us() + vblank_time;
+
+    while (wl->callback_wait) {
+
+        while (wl_display_prepare_read(wl->display) != 0)
+            wl_display_dispatch_pending(wl->display);
+        wl_display_flush(wl->display);
+
+        int poll_time = (finish_time - mp_time_us()) / 1000;
+        if (poll_time < 0)
+            poll_time = 0;
+        poll(fds, 1, poll_time);
+
+        wl_display_read_events(wl->display);
+        wl_display_dispatch_pending(wl->display);
+
+        if (finish_time - mp_time_us() <= 0)
+            break;
```

Comment:
This is not wrong, but what I meant was `if (poll_time == 0) break;`. So you don't need the second mp_time_us() call.

Reply:
Timings break if I do that though. Perhaps it's because of some delay from `wl_display_read_events` and `wl_display_dispatch_pending`?

Reply:
Actually scratch that. This display has too much latency for me to actually say that. I'll need to double check it later.

Reply:
If the duplicate check is needed, it could be put into the while condition. You could also use "a<=b" instead of "a-b<=0".
Anyway, doesn't matter too much.